feat: Implement query rules

aseure opened the pull request.

clement-leprovost commented on algoliasearch/algoliasearch.go:
```diff
@@ -313,6 +313,44 @@ type Index interface {
+	//   - `context`:     restricts the search to rules with a specific context (exact match)
+	//   - `page`:        requested page (zero-based, defaults to zero)
+	//   - `hitsPerPage`: maximum number of hits in a page (defaults to 20)
+	SearchForRules(query string, anchoring RulePatternAnchoring, context string, page, hitsPerPage int) (SearchForRulesRes, error)
```
How are you going to extend the method when we add more search parameters? Wouldn't it be safer to use a type for the parameters?

aseure replied:
I didn't know it was subject to modifications. It's already painful to pass 5 parameters. I'll make it take a map and perform a type check on the known parameters. This way, even if a new parameter is used by the user, it will let it go through.

clement-leprovost commented on algoliasearch/types_rule.go:
```diff
@@ -0,0 +1,93 @@
+	UpdatedAt string `json:"updatedAt"`
+}
+
+type SearchForRulesRes struct {
```
What happens if the API suddenly returns more fields in the response? Is it going to cause the existing code to crash/return an error? Or will the extraneous fields simply be ignored? I am concerned about forward compatibility (i.e. making deployed code work with new versions of the API).

aseure replied:
Simply ignored until we implement it in the Go client, as shown here:
https://play.golang.org/p/dSJkBWxmpw

clement-leprovost commented on algoliasearch/types_rule.go:
```diff
@@ -0,0 +1,93 @@
+	Anchoring RulePatternAnchoring `json:"anchoring"`
+	Pattern   string               `json:"pattern"`
+	Context   string               `json:"context,omitempty"`
+	Priority  int                  `json:"priority,omitempty"`
```
Sorry, maybe it wasn't obvious in the doc, but `priority` is actually not supported. It was left out of the scope until further notice. => Please remove it.

clement-leprovost commented on algoliasearch/types_rule.go:
```diff
@@ -0,0 +1,93 @@
+		Anchoring: anchoring,
+		Pattern:   pattern,
+		Context:   context,
+		Priority:  priority,
```
Same as above.

aseure commented:
@clement-leprovost My last commit should fix the concerns you had. Let me know.

aseure commented:
@clement-leprovost 

I've squashed a bit my commits, let the consequence's params be regular search query parameters and renamed `SearchForRules`/`SearchRules` to match the other API clients naming in the documentation.

Could you take a last look at this PR before I merge it?

Thank you in advance :)
